modules/mod_ginger_base/controllers/controller_search.erl

```diff
@@ -16,36 +16,89 @@ init(Args) ->
-        Results
-    ),
-    %% Serialize to JSON
-    SearchResults = #{
-        <<"result">> => [search_result(R, Context) || R <- VisibleResults],
-        <<"total">> => Total
-    },
-    Json = jsx:encode(SearchResults),
-    %% Done
-    {Json, Req, State}.
+    case proplists:get_value(mode, State) of 
+        coordinates ->
+            %% We're only interested in the geolocation and id fields
+	    Query1 = [{source, [<<"geolocation">>]} | arguments(RequestArgs)],
+            %% And it doesn't make sense to retrieve any results without coordinates
+	    Query2 = [{has_geo, <<"true">>} | Query1],
+	    SearchResults = z_search:search({Type, Query2}, {Offset + 1, Limit}, Context),
+            %% Strip any unneeded data
+	    Coordinates = 
+                lists:map(
+                  fun(Item) -> 
+                          Id = maps:get(<<"_id">>, Item),
+                          #{<<"lat">> := Lat,
+                            <<"lon">> := Lon} = 
+                              get_in([<<"_source">>, <<"geolocation">>], Item),
+                          #{id => list_to_integer(binary_to_list(Id)),
+                            lat => Lat, lng => Lon}
+                  end,
+                  SearchResults#search_result.result),
+            Result = #{<<"result">> => Coordinates,
+                       <<"total">> => SearchResults#search_result.total},
+	    Json = jiffy:encode(Result),
+	    {Json, Req, State};
+        _ ->
+	    Result = z_search:search({Type, arguments(RequestArgs)}, {Offset + 1, Limit}, Context),
+	    #search_result{
+	       result = Results,
+	       facets = _Facets,
+	       total = Total
+	      } = Result,
+	    %% Filter search results not visible for current user
+	    VisibleResults = lists:filter(
+			       fun(R) ->
+				       is_visible(R, Context)
+			       end,
+			       Results
+			      ),
+	    %% Serialize to JSON
+	    SearchResults = #{
+			      <<"result">> => [search_result(R, Context) || R <- VisibleResults],
+			      <<"total">> => Total
+			     },
+	    SR = json_map(fun(V) -> stringify_date(V, Context) end, SearchResults),
+	    Json = jiffy:encode(SR),
+	    {Json, Req, State}
+    end.
+
+%% @doc Apply a function to all leaf values of a JSON-like structure
+-type json_leaf() :: string() | binary() | integer() | true | false | null.
+-type json_node() :: map() | list() | json_leaf().
+-spec json_map(fun((json_leaf()) -> json_leaf()), json_node()) -> json_node().
+                        
+json_map(F, Elm) when is_map(Elm) ->
+    maps:fold(fun(K,V,A) -> A#{K => json_map(F, V)} end, #{}, Elm);
+json_map(F, Elm) when is_list(Elm) ->
+    lists:map(fun(V) -> json_map(F, V) end, Elm);
+json_map(F, Elm) ->
+    F(Elm).
+
+%% @doc Converts a Erlang formatted date-time to an ISO-8601 string. 
+%% Simply returns anything else
+stringify_date({{_Y, _M, _D},{_H, _Mi, _S}} = Date, Context) ->
+    z_datetime:format(Date, "c", Context);
+stringify_date(V, _Context) ->
+    V.
```

Comment:
Okay for now, but mixing concerns. Should be moved to a dedicated JSON module in the future.